icefoganalytics/internal-data-portal
# Part 1: Dataset Visualize Manage Page

Relates to:
- [0001, Data Modeling, 2024-01-05](https://docs.google.com/document/d/1Mdbl_Cy0BgKiq0DGdvOKUqJn-FdgbF0LSVykwf__FY8/edit?pli=1)
- [0002, UI Modeling, 2024-01-08](https://docs.google.com/document/d/1tP63Ob4XPBGlASPHVf1GnNb3hlIkybmWBqJDA61u3J0/edit)
- Partially completes  https://github.com/icefoganalytics/internal-data-portal/issues/54

# Context

Only builds out the visualization properties section.

## Dataset Visualize Manage Page - /dataset/:slug/visualize/manage

### Front-end routes

Various tabs at the same level as this page.

### Back-end routes

* /api/dataset-entries?where=... - filterable list of data set entry results
* /api/dataset-entries?where=....csv - filterable list of data set entry results, downloadable in csv format
* Some kind of Power BI link generator with appropriate content?

### Components

* `BaseLayout.vue` - see previous definition
* `DatasetLayout.vue` - see previous definition
* `DatasetVisualizeManagePage.vue`
    * `VisualizePropertiesFormCard.vue`
        * Allow Downloaded to CSV? - VisualizationControl#is_dowloadable_as_csv
        * Limit data view on data search - VisualizationControl#has_search_restrictions
        * Fields - VisualizationControl#has_search_field_restrictions
        * Exclude: - multi select of VisualizationControlSearchFieldExclusion#id
        * Rows - VisualizationControl#has_search_row_limits
        * (number input for row limit) - VisualizationControl#search_row_limit_maximum, has no visible label
    * `DatasetEntriesCard.vue` - to be done in a future PR

### Source Wireframe

![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/9c67a4d6-cbdc-4dd0-af2c-573ec182b885)

# Implementation

Build visualization properties form card.
Lock Visualize tab based on user policy.
Build dataset policy helpers to match dataset policy scope query functions.
Switch policy pattern to throwing a 500 if required associations are missing, instead of trying to force passing a fully functional type.
Perform saving of field exclusion, through visualization control model nested association attributes.

# Screenshots

![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/fc7aae40-4fe7-4bb9-acfd-f10c3c32bdd7)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080 as your Data Owner role'd user.
4. Create a new dataset.
5. Add some fields to the dataset via the "Fields" tab.
6. Go to the "Visualize" tab.
7. Check that all the fields auto-save correctly.
8. Check that the "fields" list shows the fields you added, and only those fields.
9. Add a Self-Serve government wide access grant.
10. Log out and log back in as low tier User role'd user.
11. Go to the datasets page, and select the dataset that you created previously.
12. Check that the Fields and Visualize tabs are locked.
13. Subscribe to the dataset, and check that the Fields and Visualize tabs unlock.
